# Part 1

Importing pandas:

In [3]:
import pandas as pd

Reading and importing data:

In [4]:
df = pd.read_csv('kc_house_data.csv')

## 1. Viewing and explaining variables

In [5]:
df.columns

Index(['id', 'date', 'price', 'bedrooms', 'bathrooms', 'sqft_living',
       'sqft_lot', 'floors', 'waterfront', 'view', 'condition', 'grade',
       'sqft_above', 'sqft_basement', 'yr_built', 'yr_renovated', 'zipcode',
       'lat', 'long', 'sqft_living15', 'sqft_lot15'],
      dtype='object')

In [6]:
df.head()

id             date     price  bedrooms  bathrooms  sqft_living  \
0  7129300520  20141013T000000  221900.0         3       1.00         1180   
1  6414100192  20141209T000000  538000.0         3       2.25         2570   
2  5631500400  20150225T000000  180000.0         2       1.00          770   
3  2487200875  20141209T000000  604000.0         4       3.00         1960   
4  1954400510  20150218T000000  510000.0         3       2.00         1680   

   sqft_lot  floors  waterfront  view  ...  grade  sqft_above  sqft_basement  \
0      5650     1.0           0     0  ...      7        1180              0   
1      7242     2.0           0     0  ...      7        2170            400   
2     10000     1.0           0     0  ...      6         770              0   
3      5000     1.0           0     0  ...      7        1050            910   
4      8080     1.0           0     0  ...      8        1680              0   

   yr_built  yr_renovated  zipcode      lat     long  sqft_living15  \
0      1955             0    98178  47.5112 -122.257           1340   
1      1951          1991    98125  47.7210 -122.319           1690   
2      1933             0    98028  47.7379 -122.233           2720   
3      1965             0    98136  47.5208 -122.393           1360   
4      1987             0    98074  47.6168 -122.045           1800   

   sqft_lot15  
0        5650  
1        7639  
2        8062  
3        5000  
4        7503  

[5 rows x 21 columns]

In [42]:
df.tail(10)

id             date      price  bedrooms  bathrooms  \
21603  7852140040  20140825T000000   507250.0         3       2.50   
21604  9834201367  20150126T000000   429000.0         3       2.00   
21605  3448900210  20141014T000000   610685.0         4       2.50   
21606  7936000429  20150326T000000  1010000.0         4       3.50   
21607  2997800021  20150219T000000   475000.0         3       2.50   
21608   263000018  20140521T000000   360000.0         3       2.50   
21609  6600060120  20150223T000000   400000.0         4       2.50   
21610  1523300141  20140623T000000   402101.0         2       0.75   
21611   291310100  20150116T000000   400000.0         3       2.50   
21612  1523300157  20141015T000000   325000.0         2       0.75   

       sqft_living  sqft_lot  floors  waterfront  view  ...  grade  \
21603         2270      5536     2.0           0     0  ...      8   
21604         1490      1126     3.0           0     0  ...      8   
21605         2520      6023     2.0           0     0  ...      9   
21606         3510      7200     2.0           0     0  ...      9   
21607         1310      1294     2.0           0     0  ...      8   
21608         1530      1131     3.0           0     0  ...      8   
21609         2310      5813     2.0           0     0  ...      8   
21610         1020      1350     2.0           0     0  ...      7   
21611         1600      2388     2.0           0     0  ...      8   
21612         1020      1076     2.0           0     0  ...      7   

       sqft_above  sqft_basement  yr_built  yr_renovated  zipcode      lat  \
21603        2270              0      2003             0    98065  47.5389   
21604        1490              0      2014             0    98144  47.5699   
21605        2520              0      2014             0    98056  47.5137   
21606        2600            910      2009             0    98136  47.5537   
21607        1180            130      2008             0    98116  47.5773   
21608        1530              0      2009             0    98103  47.6993   
21609        2310              0      2014             0    98146  47.5107   
21610        1020              0      2009             0    98144  47.5944   
21611        1600              0      2004             0    98027  47.5345   
21612        1020              0      2008             0    98144  47.5941   

          long  sqft_living15  sqft_lot15  
21603 -121.881           2270        5731  
21604 -122.288           1400        1230  
21605 -122.167           2520        6023  
21606 -122.398           2050        6200  
21607 -122.409           1330        1265  
21608 -122.346           1530        1509  
21609 -122.362           1830        7200  
21610 -122.299           1020        2007  
21611 -122.069           1410        1287  
21612 -122.299           1020        1357  

[10 rows x 21 columns]

In [9]:
df['view'].unique()

array([0, 3, 4, 2, 1], dtype=int64)

So, we're assuming that the **view** variable takes 5 values rating the view from 0 to 5, 5 being the highest.

### Variable reference:

**id**: An identifier for each sale <br>
**date**: The date when the house was sold <br>
**price**: The price of the house <br>
**bedrooms**: The number of bedrooms in each house <br>
**bathrooms**: The number of bathrooms in each house where 0.5 accounts for a bathroom attached to a room but without all the items(toilet, shower..)<br>
**sqft_living**: The living area of the house (apartment space) in square feet<br>
**sqft_lot**: The total area of the house in square feet<br>
**floors**: The number of floors <br>
**waterfront**: A variable indicating whether the house has a waterfront view; 0 if it doesn't, 1 if it does<br>
**view**: A variable that takes the values from 0 to 4 representing how good the view is<br>
**condition**: An variable that takes the values from 1 to 5 representing the condition of the house<br>
**grade**: An variable that takes the values from 1 to 13 representing the grade given to the house<br>
**sqft_above**: How many square feet the house is above the ground level (so excluding basements and underground areas)<br>
**sqft_basement**: The square footage of the house's basement<br>
**yr_built**: The year the house was built<br>
**yr_renovated**: The year the house was renovated last<br>
**zipcode**: The area zipcode where the house is in<br>
**lat**: The latitude of the house's location<br>
**long**: The longitude of the house's location<br><br>
*Side note: Latitude and longitude are parameters or coordinates that help determine the location of any place on Earth. Both latitude and longitude are angles that uniquely define points on the Earth's sphere. They constitute a coordinate system or scheme to locate or identify geographic positions anywhere on the Earth's surface.*<br>
*Source: https://www.techtarget.com/whatis/definition/latitude-and-longitude#:~:text=Latitudes%20are%20horizontal%20lines%20that,or%20places%20on%20the%20globe.*<br><br>
**sqft_living15**: The average square footage of the living space of the nearest 15 neighbors<br>
**sqft_lot15**: The average square footage of the land lots of the nearest 15 neighbors

## 2. Variable Specificities and Distribution

### Variable Types:

In [11]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21613 entries, 0 to 21612
Data columns (total 21 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   id             21613 non-null  int64  
 1   date           21613 non-null  object 
 2   price          21613 non-null  float64
 3   bedrooms       21613 non-null  int64  
 4   bathrooms      21613 non-null  float64
 5   sqft_living    21613 non-null  int64  
 6   sqft_lot       21613 non-null  int64  
 7   floors         21613 non-null  float64
 8   waterfront     21613 non-null  int64  
 9   view           21613 non-null  int64  
 10  condition      21613 non-null  int64  
 11  grade          21613 non-null  int64  
 12  sqft_above     21613 non-null  int64  
 13  sqft_basement  21613 non-null  int64  
 14  yr_built       21613 non-null  int64  
 15  yr_renovated   21613 non-null  int64  
 16  zipcode        21613 non-null  int64  
 17  lat            21613 non-null  float64
 18  long  

**Note**: The 64 in int64 refers to the memory allocated to hold this character. It represents signed integers with values that range from negative 9,223,372,036,854,775,808 through positive 9,223,372,036,854,775,807.

**Nominal Categorical**: id, zipcode *(as there is no order, and the numbers are unqiue identifiers)*<br>
**Binary Categorical**: waterfront<br>
**Ordinal Categorical**: view, condition, grade, date *(There is an order or ranking)*<br>
**Numerical (Continuous)**: price, sqft_living, sqft_lot, lat, long<br>
**Numerical (Discrete)**: bedrooms, bathrooms, floors, sqft_above, sqft_basement, yr_built, yr_renovated, sqft_living15, sqft_lot15<br>
*Note: We considered 'bathrooms' to be discrete, even though it takes decimal values, and that's assuming it only takes .25, .5, .75*<br>
*Same goes for floors, assuming there are attics in some houses, thus the .5* 

Note that some variables are not regarded as numerical even though their Dtype is int64 (example: view).

### Common values

One way to find the number of occurences of a value is using the method ```value_counts()```:

In [12]:
df.columns

Index(['id', 'date', 'price', 'bedrooms', 'bathrooms', 'sqft_living',
       'sqft_lot', 'floors', 'waterfront', 'view', 'condition', 'grade',
       'sqft_above', 'sqft_basement', 'yr_built', 'yr_renovated', 'zipcode',
       'lat', 'long', 'sqft_living15', 'sqft_lot15'],
      dtype='object')

In [13]:
# Most observed number of floors:
df['floors'].value_counts()

1.0    10680
2.0     8241
1.5     1910
3.0      613
2.5      161
3.5        8
Name: floors, dtype: int64

The decimal part is explained in the above section.

In [14]:
df.waterfront.value_counts()

0    21450
1      163
Name: waterfront, dtype: int64

#### Some insights:

In [31]:
no_water_view = df[df['waterfront'] == 0]
water_view = df[df['waterfront'] == 1]

In [33]:
print(water_view['price'].mean().round(2),no_water_view['price'].mean().round(2))

1662524.18 531653.41


The average price of the houses with no waterfront view (531,653.41) is much smaller than that of houses with waterfront view (1,662,524.18). This potentially indicate a correlation between the two variables 'waterfront' and 'view'.

### Max and Min values:

Let's check the maximum and minimum values of some of our variables:

In [34]:
df['bedrooms'].max()

33

In [38]:
df[df['bedrooms']==33]

id             date     price  bedrooms  bathrooms  \
15870  2402100895  20140625T000000  640000.0        33       1.75   

       sqft_living  sqft_lot  floors  waterfront  view  ...  grade  \
15870         1620      6000     1.0           0     0  ...      7   

       sqft_above  sqft_basement  yr_built  yr_renovated  zipcode      lat  \
15870        1040            580      1947             0    98103  47.6878   

          long  sqft_living15  sqft_lot15  
15870 -122.331           1330        4700  

[1 rows x 21 columns]

A house with 33 bedrooms has 1 bathroom other than the one exclusive to one of the bedrooms, something seems off...

In [39]:
df.nlargest(5,'price')

id             date      price  bedrooms  bathrooms  \
7252  6762700020  20141013T000000  7700000.0         6       8.00   
3914  9808700762  20140611T000000  7060000.0         5       4.50   
9254  9208900037  20140919T000000  6890000.0         6       7.75   
4411  2470100110  20140804T000000  5570000.0         5       5.75   
1448  8907500070  20150413T000000  5350000.0         5       5.00   

      sqft_living  sqft_lot  floors  waterfront  view  ...  grade  sqft_above  \
7252        12050     27600     2.5           0     3  ...     13        8570   
3914        10040     37325     2.0           1     2  ...     11        7680   
9254         9890     31374     2.0           0     4  ...     13        8860   
4411         9200     35069     2.0           0     0  ...     13        6200   
1448         8000     23985     2.0           0     4  ...     12        6720   

      sqft_basement  yr_built  yr_renovated  zipcode      lat     long  \
7252           3480      1910          1987    98102  47.6298 -122.323   
3914           2360      1940          2001    98004  47.6500 -122.214   
9254           1030      2001             0    98039  47.6305 -122.240   
4411           3000      2001             0    98039  47.6289 -122.233   
1448           1280      2009             0    98004  47.6232 -122.220   

      sqft_living15  sqft_lot15  
7252           3940        8800  
3914           3930       25449  
9254           4540       42730  
4411           3560       24345  
1448           4600       21750  

[5 rows x 21 columns]

The priciest house has 6 bedrooms, much less than 33. It has more bathrooms, a bigger area, a better view, and is graded higher. This could mean that there was a mistake while recording the data, that entry might be considered an outlier.

In [41]:
df.nsmallest(10,'sqft_living')

id             date     price  bedrooms  bathrooms  \
19452  3980300371  20140926T000000  142000.0         0       0.00   
15381  2856101479  20140701T000000  276000.0         1       0.75   
860    1723049033  20140620T000000  245000.0         1       0.75   
18379  1222029077  20141029T000000  265000.0         0       0.75   
4868   6896300380  20141002T000000  228000.0         0       1.00   
21332  9266700190  20150511T000000  245000.0         1       1.00   
8623   6303400395  20150130T000000  325000.0         1       0.75   
11500  4322200105  20150331T000000  229050.0         1       1.00   
14466  7549801385  20140612T000000  280000.0         1       0.75   
465    8658300340  20140523T000000   80000.0         1       0.75   

       sqft_living  sqft_lot  floors  waterfront  view  ...  grade  \
19452          290     20875     1.0           0     0  ...      1   
15381          370      1801     1.0           0     0  ...      5   
860            380     15000     1.0           0     0  ...      5   
18379          384    213444     1.0           0     0  ...      4   
4868           390      5900     1.0           0     0  ...      4   
21332          390      2000     1.0           0     0  ...      6   
8623           410      8636     1.0           0     0  ...      4   
11500          420      3298     1.0           0     0  ...      4   
14466          420      6720     1.0           0     0  ...      5   
465            430      5050     1.0           0     0  ...      4   

       sqft_above  sqft_basement  yr_built  yr_renovated  zipcode      lat  \
19452         290              0      1963             0    98024  47.5308   
15381         370              0      1923             0    98117  47.6778   
860           380              0      1963             0    98168  47.4810   
18379         384              0      2003             0    98070  47.4177   
4868          390              0      1953             0    98118  47.5260   
21332         390              0      1920             0    98103  47.6938   
8623          410              0      1953             0    98146  47.5077   
11500         420              0      1949             0    98136  47.5375   
14466         420              0      1922             0    98108  47.5520   
465           430              0      1912             0    98014  47.6499   

          long  sqft_living15  sqft_lot15  
19452 -121.888           1620       22850  
15381 -122.389           1340        5000  
860   -122.323           1170       15000  
18379 -122.491           1920      224341  
4868  -122.261           2170        6000  
21332 -122.347           1340        5100  
8623  -122.357           1190        8636  
11500 -122.391           1460        4975  
14466 -122.311           1420        6720  
465   -121.909           1200        7500  

[10 rows x 21 columns]

If we look at the price, sqft_living, and the sqft_lot variables, we could notice some kind of relation between the three. So, one might affect the other; could be worth looking into and drawing a conclusion.

## 3.  Time, Latitude, and Longitude

In [45]:
df.columns

Index(['id', 'date', 'price', 'bedrooms', 'bathrooms', 'sqft_living',
       'sqft_lot', 'floors', 'waterfront', 'view', 'condition', 'grade',
       'sqft_above', 'sqft_basement', 'yr_built', 'yr_renovated', 'zipcode',
       'lat', 'long', 'sqft_living15', 'sqft_lot15'],
      dtype='object')

In [61]:
top_prices = df.nlargest(30,'price')
top_prices[['date','lat','long','price','grade','sqft_lot']]

date      lat     long      price  grade  sqft_lot
7252   20141013T000000  47.6298 -122.323  7700000.0     13     27600
3914   20140611T000000  47.6500 -122.214  7060000.0     11     37325
9254   20140919T000000  47.6305 -122.240  6890000.0     13     31374
4411   20140804T000000  47.6289 -122.233  5570000.0     13     35069
1448   20150413T000000  47.6232 -122.220  5350000.0     12     23985
1315   20150413T000000  47.5631 -122.210  5300000.0     12     24829
1164   20141020T000000  47.6767 -122.211  5110000.0     12     45517
8092   20140617T000000  47.5570 -122.210  4670000.0     12     13068
2626   20140815T000000  47.7493 -122.280  4500000.0     12     40014
8638   20140618T000000  47.6208 -122.219  4490000.0     12     27517
12370  20150506T000000  47.5692 -122.189  4210000.0     12     21540
4149   20141014T000000  47.6151 -122.224  4000000.0     12     16573
2085   20141114T000000  47.5850 -122.222  3850000.0     11     21300
7035   20140701T000000  47.6229 -122.220  3800000.0     13     42840
19017  20140911T000000  47.7296 -122.370  3800000.0     13     35000
16302  20140530T000000  47.6395 -122.234  3710000.0     12     28078
6508   20150421T000000  47.6515 -122.277  3650000.0     11     19401
18482  20150511T000000  47.6146 -122.213  3650000.0     12      8694
15255  20140911T000000  47.6409 -122.241  3640000.0     11     22257
19148  20141011T000000  47.6165 -122.236  3640000.0     12     19897
2864   20141118T000000  47.5925 -122.287  3600000.0     10     12431
21576  20150107T000000  47.5943 -122.110  3570000.0     10     10584
5880   20141007T000000  47.6209 -122.237  3420000.0     11     20412
7989   20140623T000000  47.5348 -122.243  3400000.0     10     17826
10446  20150508T000000  47.6380 -122.288  3400000.0     11     11765
13967  20150303T000000  47.6967 -122.216  3400000.0     10     47870
20460  20150402T000000  47.6480 -122.218  3350000.0     11     15360
14083  20140904T000000  47.6050 -122.112  3300000.0     11     13873
18199  20141230T000000  47.6083 -122.110  3300000.0     11     41300
18302  20140702T000000  47.5675 -122.189  3300000.0     11     21738

It seems a bit difficult to spot a relation between the date and prices, further analysis should be performed. As for the latitude and longitude, I'm not sure what to conclude based on the numbers, as they seem close to each other, generally.